File: lectures/13/Books/app/src/main/java/ro/ubbcluj/cs/books/Manager.java
package ro.ubbcluj.cs.books;


import android.app.Application;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.view.View;
import android.widget.ProgressBar;

import java.util.List;

import ro.ubbcluj.cs.books.domain.Book;
import ro.ubbcluj.cs.books.service.BookService;
import ro.ubbcluj.cs.books.service.ServiceFactory;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import timber.log.Timber;

class Manager {
  private BookApp app;
  private BookService service;

  Manager(Application application) {
    this.app = (BookApp) application;
    service = ServiceFactory.createRetrofitService(BookService.class,
        BookService.SERVICE_ENDPOINT);
  }


  boolean networkConnectivity(Context context) {
    ConnectivityManager cm = (ConnectivityManager) context
        .getSystemService(Context.CONNECTIVITY_SERVICE);
    assert cm != null;
    NetworkInfo networkInfo = cm.getActiveNetworkInfo();
    return networkInfo != null && networkInfo.isConnected();
  }

  void loadEvents(final ProgressBar progressBar, final MyCallback callback) {

    service.getBooks()
        .subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new Subscriber<List<Book>>() {
          @Override
          public void onCompleted() {
            Timber.v("Book Service completed");
            progressBar.setVisibility(View.GONE);
          }

          @Override
          public void onError(Throwable e) {
            Timber.e(e, "Error while loading the events");
            callback.showError("Not able to retrieve the data. Displaying local data!");
          }

          @Override
          public void onNext(final List<Book> books) {
            new Thread(new Runnable() {
              public void run() {
                app.db.getBookDao().deleteBooks();
                app.db.getBookDao().addBooks(books);
              }
            }).start();
            Timber.v("Books persisted");
          }
        });
  }

  void save(final Book book, final MyCallback callback) {
    service.addBook(book)
        .subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new Subscriber<Book>() {
          @Override
          public void onCompleted() {
            addDataLocally(book);
            Timber.v("Book Service completed");
            callback.clear();
          }

          @Override
          public void onError(Throwable e) {
            Timber.e(e, "Error while persisting an book");
            callback.showError("Not able to connect to the server, will not persist!");
          }

          @Override
          public void onNext(Book book) {
            Timber.v("Book persisted");
          }
        });
  }

  private void addDataLocally(final Book book) {
    new Thread(new Runnable() {
      public void run() {
        app.db.getBookDao().addBook(book);
      }
    }).start();
  }
}

File: archives/2018-2019/13/Books/app/src/main/java/ro/ubbcluj/cs/books/Manager.kt
package ro.ubbcluj.cs.books


import android.content.Context
import android.net.ConnectivityManager
import android.view.View
import android.widget.ProgressBar

import ro.ubbcluj.cs.books.domain.Book
import ro.ubbcluj.cs.books.service.BookService
import ro.ubbcluj.cs.books.service.ServiceFactory
import ro.ubbcluj.cs.books.utils.logd
import ro.ubbcluj.cs.books.utils.loge
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

internal class Manager(val app: BookApp) {
  private val service: BookService = ServiceFactory
      .createRetrofitService(BookService::class.java, BookService.SERVICE_ENDPOINT)


  fun networkConnectivity(context: Context): Boolean {
    val cm = context
        .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    val networkInfo = cm.activeNetworkInfo
    return networkInfo != null && networkInfo.isConnected
  }

  fun loadEvents(progressBar: ProgressBar, callback: MyCallback) {

    service.books
        .subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(object : Subscriber<List<Book>>() {
          override fun onCompleted() {
            logd("Book Service completed")
            progressBar.visibility = View.GONE
          }

          override fun onError(e: Throwable) {
            loge("Error while loading the events", e)
            callback.showError("Not able to retrieve the data. Displaying local data!")
          }

          override fun onNext(books: List<Book>) {
            Thread(Runnable {
              app.db.bookDao.deleteBooks()
              app.db.bookDao.addBooks(books)
            }).start()
            logd("Books persisted")
          }
        })
  }

  fun save(book: Book, callback: MyCallback) {
    service.addBook(book)
        .subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(object : Subscriber<Book>() {
          override fun onCompleted() {
            addDataLocally(book)
            logd("Book Service completed")
            callback.clear()
          }

          override fun onError(e: Throwable) {
            loge("Error while persisting an book", e)
            callback.showError("Not able to connect to the server, will not persist!")
          }

          override fun onNext(book: Book) {
            logd("Book persisted")
          }
        })
  }

  private fun addDataLocally(book: Book) {
    Thread(Runnable { app.db.bookDao.addBook(book) }).start()
  }
}
